### ISSUE TITLE
Tracking and Improving Exception / Error Messages in Microsoft Azure SDK Libraries

### ISSUE THREAD

**Issue by JonathanGiles** [Azure/azure-sdk-for-java#23038]
## Problem Description / Background

This project recognises that developers face numerous difficulties when using libraries as part of their software development process. From poor documentation and lackluster samples to ill-considered API design, using libraries is not free of friction. At Microsoft, our goal is to reduce this friction wherever possible, to delight our end users and to provide them with the most productive development experience possible. It is important to note that friction doesn't exist on the ideal path, it also exists on the failure paths. When things go wrong, developers are forced to turn to Google, to Stack Overflow, to forums and blog posts. They copy and past their error messages and hope to find someone who has already suffered from the same failure.

One avenue of investigation currently is in improving our runtime feedback to developers when things go wrong. Whenever a user reports an issue to us, oftentimes the error message is computer-generated and referring to some category of problem - a null pointer exception, a class not found exception, an out of memory exception, and so on. The ability for the user to reason about this error message and do something appropriate is often too complex of an ask for them.

Ideally we would intercept many of these failure messages before the user sees them, and rewrite them in plain English, with clear descriptions of how to resolve the issue, or a link to an online article published by the engineering team. 

What we need is a way to capture error output from our libraries, and to track this output over time to ensure we can clearly see what needs improving, and what has already been improved. Essentially, if we can create a form of unit test for our libraries, we could then collect the output and log it to a central web app where users can navigate to API they care about and identify areas where the tool has found our error output to be wanting.

An extension to this tool would be to generate and then run randomised tests with known edge cases (null arguments, large values, etc) to help capture other failure conditions and to help seed the set of unit tests until users can write more of these tests of how to resolve the issue, or a link to an online article published by the engineering team. 

## Scope

The end output from this project will exist of the following:

1. A Java-based framework (similar in nature to JUnit) that can execute all methods annotated appropriately, gather any console / log output for a particular method call, and then, at the end of the execution of all tests, generate a language agnostic (i.e. JSON) representation of the results.
2. A web-app that acts as both a receiver of the test output, and as a displayer of the results.
   1. The tool output, as JSON, will need to be submitted to an endpoint offered by the web app in a secure fashion.
   2. The web app will need to parse the JSON output, determine the language it represents (as whilst we will start with Java, support for other languages will eventually be desirable), and to either create a new library representation or append the results to a previous run of the library execution.
   3. The front end of the web app will allow users to browse through all languages and all libraries, and see the output for each test run. By default the latest result may be shown for each test run, and there must be a way for users to mark if the output has been reviewed as `acceptable`, `won't fix`, `needs fix`, and so on. Based on the review state, colour coding can be used to visualise the overall state of the library. As new executions of the tool are submitted, the web app must compare the previous result to the new one for each test, and for each test that is unchanged, the review should remain, otherwise a 'needs review' state should be entered into.

## Requirements

1. Make use of the Microsoft Azure cloud wherever necessary. Research your requirements and ask for the functionality you require, e.g. Azure Cosmos DB or Azure Virtual Machines.

## Deliverables

1. Production-ready exception / error gathering framework.
   1. In the first instance the goal is to support Java, but as a stretch goal support for C#, Python, TypeScript, or other languages can be implemented as well.
   2. Another stretch goal is to generate a base-level of error tests by inspecting public API in a given library and passing in good test inputs, e.g. null for objects, negative values, extremely large values, etc.
2. Production-ready webapp for receipt, display, and review of error messages and exceptions.
   1. Web-app has a high-quality user interface that is intuitive and makes the process of error review and tracking over time simple.

**Comment by SivaTharun** [comment]
@JonathanGiles can i work on this issue?

**Comment by JonathanGiles** [comment]
Hi @SivaTharun - thank you for your interest in helping us out! I was just in a call yesterday with three students who are wanting to work on this as part of their university studies. Shortly I was going to assign this issue to them, but I believe their work on this will be open source and you are welcome to contribute to it. Thanks again!

**Comment by JonathanGiles** [comment]
The three people who have teamed up to work on this are @MattGerrand, @dashdotme, and @devoiddroid.
